### PR TITLE
Enabling --non-masquerade-cidr option

### LINKED ISSUE #1360: k8s integration with existing private network
**Is this a request for help?**:
---
No

**Is this an ISSUE or FEATURE REQUEST?** (choose one):
---
Feature Request

**What version of acs-engine?**:
---
0.5.0

**Orchestrator and version (e.g. Kubernetes, DC/OS, Swarm)**

Kubernetes 1.6

**What happened**:

We want to integrate a k8s cluster into our private network linked to Azure using a VPN. 
We made a successful deployment using and existing VNET which address space is 10.A.B.0/26
However we ran into some issues with the masquerading exclusion of the traffic which is by default set on 10.0.0.0/8. 
Network traffic coming from the pods could not be routed back because their sources address where not known outside of the cluster. 

**What you expected to happen**:

We want to have the network traffic properly routed from the pods to the outside of the cluster and back. 

There is a configuration option at the kubelet level (--non-masquerade-cidr option).

So we want to submit a fix for the ACS Engine, and will listen to any advice on it or any related subject in order to have a robust existing subnet integration in a private network context.

**How to reproduce it** (as minimally and precisely as possible):

Set up a k8s deployment using an existing VNET integrated in a private network overlapping with the 10.0.0.0/8 space. 

**Anything else we need to know**:

...

### PR DESCRIPTION
**Pull request by dbourasseau**
**What this PR does / why we need it**:

This PR enables configuration of the --non-masquerade-cidr option at the kubelet level.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #1360

**Special notes for your reviewer**:

**Release note**:

```release-note NONE
```

---
This change is [Reviewable](https://reviewable.io/reviews/azure/acs-engine/1361)

### CODE REVIEW

File: parts/kuberneteskubelet.service [hunk 1]
```diff
@@ -50,6 +50,7 @@ ExecStart=/usr/bin/docker run \
         --network-plugin=${KUBELET_NETWORK_PLUGIN} \
         --max-pods=${KUBELET_MAX_PODS} \
         --node-status-update-frequency=${KUBELET_NODE_STATUS_UPDATE_FREQUENCY} \
+        --non-masquerade-cidr=${KUBELET_NON_MASQUERADE_CIDR} \
```
Comment (lachie83) [inline parts/kuberneteskubelet.service]: I don't see this as a valid option on the documentation. Can you confirm? -- https://kubernetes.io/docs/admin/kubelet/

File: parts/kubernetesagentcustomdata.yml [hunk 2]
```diff
@@ -115,6 +115,7 @@ write_files:
     KUBE_CTRL_MGR_POD_EVICTION_TIMEOUT={{WrapAsVariable "kubernetesCtrlMgrPodEvictionTimeout"}}
     KUBE_CTRL_MGR_ROUTE_RECONCILIATION_PERIOD={{WrapAsVariable "kubernetesCtrlMgrRouteReconciliationPeriod"}}
 {{if IsKubernetesVersionGe "1.6.0"}}
+    KUBELET_NON_MASQUERADE_CIDR={{WrapAsVariable "kubernetesNonMasqueradeCIDR"}}
```
Comment (lachie83) [inline parts/kubernetesagentcustomdata.yml]: Depending on what the option can take. Should this be an array of CIDR blocks?

File: parts/kuberneteskubelet.service (hunk 1 above)
Comment (SaaldjorMike) [reply parts/kuberneteskubelet.service]: This page at least mentions it: https://kubernetes.io/docs/concepts/cluster-administration/network-plugins/ 

Pretty sure I saw it somewhere on the kubelet options previously.

**Comment by dbourasseau** [comment]
Any news ? Have you watched the pull request? must we do something else?

**Comment by mrosterm** [comment]
Whats the status on this one?

**Comment by dbourasseau** [comment]
We are still waiting for your response

**Comment by SaaldjorMike** [comment]
Linking this with issue #425

**Comment by SaaldjorMike** [comment]
According to https://github.com/kubernetes/kubernetes/blob/master/cmd/kubelet/app/options/options.go#L339 `--non-masquerade-cidr` is deprecated.

**Comment by jackfrancis** [comment]
Closing this as `--non-masquerade-cidr` is deprecated. Happy to iterate on a more current solution if you'd like to file an issue that describes the high order intention.

**Comment by bathizte** [comment]
Hello @jackfrancis 

The high order intention is hopefully explained in #1360 linked issue. 

We understand the option has been deprecated a month ago in master and will be removed in a future version but : 
- K8s deployments using pre-1.8 version still need some parts of this PR if deploying on a private subnet (unless we have missed something).
- We may want to check that along with deprecating the option, a fix or an other mechanism has been commted to deploy a k8s cluster in a private network and disable masquerading on the pod cidr. 
In the case a new option allows this, let's update the PR specifying the new option for Acs engine k8s v>=1.8 deployments.

**Comment by jackfrancis** [comment]
@bathizte We did some further investigation, and this is still the only way (as far as we know) to do this in v1.8, so we'll proceed. Thanks!

**Comment by sylr** [comment]
I had a similar PR for this (https://github.com/Azure/acs-engine/pull/1142) but I closed it because if I think I saw that the --non-masquerade-cidr option of kubelet is deprecated.

**Comment by lachie83** [comment]
@jackfrancis when can we get this merged?

**Comment by jackfrancis** [comment]
Been having trouble getting it to pass Windows e2e, will re-poke.

**Comment by karataliu** [comment]
We'd better begin seeking for new solutions, as @sylr mentioned this flag is deprecated:

https://github.com/kubernetes/kubernetes/blob/v1.8.0/cmd/kubelet/app/options/options.go#L340

**Comment by tanner-bruce** [comment]
@jackfrancis @karataliu I just came across the [ip-masq-agent](https://github.com/kubernetes-incubator/ip-masq-agent) today

**Pull request merged**